subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -60,24 +64,59 @@
 public class GroovyCompile extends AbstractCompile {
     private Compiler<GroovyJavaJointCompileSpec> compiler;
     private FileCollection groovyClasspath;
+    private ConfigurableFileCollection compilerPluginClasspath;
     private final CompileOptions compileOptions;
     private final GroovyCompileOptions groovyCompileOptions = new GroovyCompileOptions();
 
     public GroovyCompile() {
-        CompileOptions compileOptions = getServices().get(ObjectFactory.class).newInstance(CompileOptions.class);
+        ObjectFactory objectFactory = getServices().get(ObjectFactory.class);
+        CompileOptions compileOptions = objectFactory.newInstance(CompileOptions.class);
         this.compileOptions = compileOptions;
+        this.compilerPluginClasspath = objectFactory.fileCollection();
         CompilerForkUtils.doNotCacheIfForkingViaExecutable(compileOptions, getOutputs());
     }
 
+    @Override
+    @CompileClasspath
+    public FileCollection getClasspath() {
+        // Note that this is an approximation and must be fixed before de-incubating getCompilerPluginClasspath()
+        // See https://github.com/gradle/gradle/pull/9513
+        return super.getClasspath();
+    }
+
+    /**
+     * The classpath for compiler plugin to run, for example, AST transformation.
+     *
+     * @since 5.6
+     */
+    @Classpath
+    @Incubating
+    public ConfigurableFileCollection getCompilerPluginClasspath() {
+        return compilerPluginClasspath;
+    }
+
+    private boolean experimentalCompilationAvoidanceEnabled() {
+        return Boolean.getBoolean("org.gradle.groovy.compilation.avoidance");
+    }
+
     @Override
     @TaskAction
     protected void compile() {
         checkGroovyClasspathIsNonEmpty();
+        separateCompileClasspathIfCompileAvoidanceEnabled();
```

Comment:
❌ The separation happens too late. This is when the task execution happens, input fingerprinting already finished at this point. The logic needs to be part of the `compilerPluginClasspath`'s getter.

Why don't you use the same approach as I did here? https://github.com/gradle/gradle/pull/9388/files#diff-8b2630613213a6518a3beff2cc35ad47R121

This also means that we need a test that shows that compile avoidance is disabled when the flag is not set.

Reply:
Added a test in https://github.com/gradle/gradle/blob/ffbfc18b7aec5fff9536e0c4b1280b322695180d/subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy#L95